### PR TITLE
DM-6287 Charts refactoring

### PR DESCRIPTION
**Pull request by tgoldina**
- Moved chart-related code to charts package
- Converted React components to ES6 classes
- Reorganized controllers to store chart data (xyplot and histogram) under charts space

Fixed bugs 
- missing chart mount action, when a chart is removed and then recreated on the same table
- undefined shows as a label when no server call is necessary

### CODE REVIEW

File: src/firefly/js/charts/ChartsCntlr.js [hunk 1]
```diff
@@ -75,44 +151,39 @@ export function reducer(state={ui:{}}, action={}) {
...
                     }
                 }
                 return state;
             case (DELETE) :
-                if (has(state, ['tbl', tblId, uChartId])) {
-                    if (Object.keys(state['tbl'][tblId]).length > 1) {
-                        return updateSet(state, ['tbl', tblId], omit(state['tbl'][tblId], [uChartId]));
+                if (has(state, [tblId, uChartId])) {
+                    if (Object.keys(state[tblId]).length > 1) {
+                        return updateSet(state, tblId, omit(state[tblId], [uChartId]));
```
Comment (loitly) [inline src/firefly/js/charts/ChartsCntlr.js]: WebUtil.updateDelete may do the same thing.

File: src/firefly/js/charts/ChartsCntlr.js [hunk 2]
```diff
@@ -75,44 +151,39 @@ export function reducer(state={ui:{}}, action={}) {
...
                     } else {
-                        return updateSet(state, 'tbl', omit(state['tbl'], [tblId]));
+                        return Object.assign({}, omit(state, [tblId]));
```
Comment (loitly) [inline src/firefly/js/charts/ChartsCntlr.js]: omit returns a new object.. assign is not necessary.

File: src/firefly/js/charts/XYPlotCntlr.js [hunk 3]
```diff
@@ -366,18 +372,24 @@ function getUpdatedParams(xyPlotParams, tableModel) {
...
     if (!get(xyPlotParams, 'x.unit')) {
-        const xUnit = get(tableModel, 'tableData.columns.0.units');
-        if (xUnit) {
-            newParams = updateSet(newParams, 'x.unit', xUnit);
+        const xIdx = getColumnIdx(tableModel, get(xyPlotParams, 'x.columnOrExpr'));
+        if (xIdx >= 0 ) {
+            const xUnit = get(tableModel, `tableData.columns.${xIdx}.units`);
+            if (xUnit) {
+                newParams = updateSet(newParams, 'x.unit', xUnit);
+            }
```
Comment (loitly) [inline src/firefly/js/charts/XYPlotCntlr.js]: there should be a util function that returns the column object given tableModel and colName.  this code here will make it hard to refactor later on.  if you don't mine adding one, i would appreciate it.
`function getColumn(tableModel, colName)`

File: src/firefly/js/charts/ui/Histogram.jsx [hunk 4]
```diff
@@ -0,0 +1,389 @@
...
+                const yReversed = (nreversed && nreversed.indexOf('y')>-1 ? true : false);
+                const xReversed = (nreversed && nreversed.indexOf('x')>-1 ? true : false);
+                chart.xAxis[0].update({'reversed' : xReversed}, false);
+                chart.yAxis[0].update({'reversed' : yReversed}, false);
+                doUpdate = true;
+            }
+            const nreversed = nextProps.reversed;
+            if (reversed !== nreversed){
+                const yReversed = (nreversed && nreversed.indexOf('y')>-1 ? true : false);
+                const xReversed = (nreversed && nreversed.indexOf('x')>-1 ? true : false);
```
Comment (loitly) [inline src/firefly/js/charts/ui/Histogram.jsx]: alternative.. not necessarily better:  !!(nreversed && nreversed.indexOf('x')>-1)

**Comment by loitly** [comment]
suggestion for xyplot optons panel:
  move 'apply', 'clear', and 'reset' to the top and make it not scrollable.  as is, in demo page where the buttons are cleanly hidden, it's a bit confusing what to do next after you've enter the options.

**Comment by loitly** [comment]
I did minimal spot testing and it looks good.

**Comment by ejoliet** [comment]
Randomly checking, i've spotted an issue:
- controls (selection, filter) works but zoom doesn't work in firefox or safari (works in chrome though)

**Comment by ejoliet** [comment]
Another small one: 'Reset' doesn't reset anything - what is 'reset' suppose to do?

**Comment by ejoliet** [comment]
Last one, related to top container bar, maybe less important because it appears in the playground environment but anyway: 
When i open the 'Charts' dialog panel and select a column from 'Col' link, clicking 'OK' made the 'Charts' dialog to close when i expect to continue and select the other column and only then i can 'Apply' my selections and close the dialog to show the updated plot. What do you think?

**Comment by xiuqin** [comment]
I agree with Emmanuel on this one. 
It should behave the same as the option pop-up when clicking on the ‘gear’ icon.

Xiuqin

**Comment by xiuqin** [comment]
zoom does not work in my Chrome browser.

Xiuqin

**Comment by xiuqin** [comment]
This is a big change from the current OPS version. 
Lets hold off until we hear from the users.

Xiuqin

**Comment by tgoldina** [comment]
Reset is resetting the form values to match current parameters of the
current plot. Clear clears the values to some reasonable default so that
you can define a new chart.

Hopefully, I have fixed the zoom issue.

2016-07-14 11:31 GMT-07:00 Emmanuel Joliet <notifications@github.com>:

**Comment by tgoldina** [comment]
This one is a bug, related to table actions affecting layout. The layout is
being changed on TABLE_REMOVE. It has the same nature as expanded layout
being changed to unexpanded on table filter or sort. I have reported it in
DM-6909.

2016-07-14 13:21 GMT-07:00 Emmanuel Joliet <notifications@github.com>:

**Comment by tgoldina** [comment]
Hopefully, I have fixed zoom.

I have put the buttons on top for convenience, but they will scroll
away. It's only in side options and easy to change back. I could not figure
out the style to scroll just the form.

2016-07-14 15:53 GMT-07:00 Tatiana Goldina <tgoldina@gmail.com>:

**Pull request merged**